### PR TITLE
Refactor disabling moves, other miscellaneous bug fixes

### PR DESCRIPTION
**Pull request by ascriptmaster**
The intention for this event is to remove the move disabling code away
from ModifyPokemon and to an event that can be run a fewer number of
times. Since the disabledMoves index is used to gray-out moves that cannot
be used, and not for any other purpose, there is no need for the related
code to be run at the ModifyPokemon timing, instead working better as a
once per turn event.

-----

This is another one of those giant refactors, so naturally I'll be nervous about
it no matter how well it seems to go.

Code ostensibly works, and passes the regressions I set up, but I don't have 
a full understanding of all the mechanics for moves, and the oldgens 
regressions are non-existent right now. Tagging @Joimer for his expertise
in all things oldgens. Also @Slayer95 for current gen stuff (especially Imprison)

Also, Bulbapedia reports that Heal Block stops Water Absorb entirely, so can
I have @Marty-D double-check and verify that? The tests I have relating to
that are being skipped right now.

@Zarel because I don't really have any idea on how NEXT Telepathy works.

### CODE REVIEW

**Comment by Marty-D** [comment]
Heal Block doesn't stop Water Absorb, quite the opposite. In Gen 4, Heal Block only stops moves from having healing effects; Abilities and items still work as usual. In Gen 5 and 6 no healing of any kind is allowed unless you count Pain Split and Regenerator, not even if you Bug Bite or Fling a healing Berry. Water Absorb and other such effects continue to have the immunity work, they just don't heal.

Minor nitpick in the Assault Vest test - it doesn't prevent the execution of Status moves, just the selection, similarly to Torment.

Otherwise, looks good on first glance. Nice tests! :+1:

**Comment by Slayer95** [comment]
Great!

Probably <code>DisableMove</code> should be ran <i>after</i> Triples' automatic center? (Is there even a move for which that would matter?)

File: test/simulator/moves/imprison.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+	afterEach(function () {
+		battle.destroy();
+	});
+
+	it('should prevent foes from using moves that the user knows', function () {
+		battle = BattleEngine.Battle.construct();
+		battle.join('p1', 'Guest 1', 1, [{species: 'Abra', ability: 'synchronize', moves: ['imprison', 'teleport']}]);
+		battle.join('p2', 'Guest 2', 1, [{species: 'Abra', ability: 'synchronize', moves: ['teleport', 'confusion']}]);
+		battle.commitDecisions();
+		battle.commitDecisions();
```
Comment (Slayer95) [inline test/simulator/moves/imprison.js]: We could save a turn and gain readibility if the first Abra had Prankster or the later had Lagging Tail.

File: test/simulator/moves/taunt.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+		battle.commitDecisions();
+		assert.strictEqual(battle.p2.active[0].boosts['spa'], 0);
+		assert.strictEqual(battle.p2.active[0].boosts['spd'], 0);
+	});
+
+	it('should disable the use of Status moves', function () {
+		battle = BattleEngine.Battle.construct();
+		battle.join('p1', 'Guest 1', 1, [{species: 'Sableye', ability: 'prankster', moves: ['taunt']}]);
+		battle.join('p2', 'Guest 2', 1, [{species: 'Spiritomb', ability: 'pressure', moves: ['calmmind']}]);
+		battle.commitDecisions();
```
Comment (Slayer95) [inline test/simulator/moves/taunt.js]: Something here to tell that Spiritomb couldn't perform CM would be nice

File: test/simulator/moves/imprison.js (hunk 1 above)
Comment (ascriptmaster) [inline test/simulator/moves/imprison.js]: Not for Imprison it won't. Checking whether the Teleport is disabled requires the extra turn anyways.

File: test/simulator/moves/taunt.js (hunk 2 above)
Comment (ascriptmaster) [inline test/simulator/moves/taunt.js]: It was done in the test above, but I agree that the tests should be merged.

**Comment by ascriptmaster** [comment]
Updated tests, or something.

I probably need more Gravity tests, what should I check for besides grounding and preventing airborne moves?

**Comment by ascriptmaster** [comment]
As for the auto-center thing, there are no moves or abilities that only target adjacent foes only for disabling, either working on all foes (Imprison) or one target (Disable), to my knowledge.

**Comment by Zarel** [comment]
Looks good to me.

**Comment by Marty-D** [comment]
For the Heal Block test with Leech Seed, Leech Seed could miss and have the test fail.

If you want more Gravity tests, you could do one for Flying types/Levitate switching into Spikes or something. Doesn't seem that necessary, but might be useful for a future refactor of airborne-related things.

**Comment by ascriptmaster** [comment]
Upgraded tests for Heal Block so much that two things broke. @Marty-D should verify that the things I "fixed" actually need fixing.

**Comment by ascriptmaster** [comment]
I can see that the `DisableMove` fix I put in March Seasonals made it merge conflict because it doesn't exist any more, so I'll fix that when I get home. For now, does the rest of the code look good?

**Comment by Marty-D** [comment]
The new Gen 4 Heal Block-related stuff looks good to me; nice fixes on the Berry healing issues too. Well done!

**Comment by ascriptmaster** [comment]
Oh, just to make sure, does the residual effect of Wish get blocked by Gen 4 Heal Block?

**Comment by Marty-D** [comment]
It's extremely difficult to have happen since the Wish user needs to be faster than the Heal Block user, then Baton Pass it out, but yes it's blocked in all gens.

**Comment by ascriptmaster** [comment]
Merge conflicts fixed. Also Gen 4 Heal Block-Wish is fixed.

**Pull request merged**